crates/sequencer/src/block_template/block_assembly.rs

```diff
@@ -251,54 +252,54 @@ fn prepare_l1_segment(
     }
 }
 
-/// Check if block has the checkpoint we are expecting and checkpoint is valid.
+/// Check if ASM manifest has the checkpoint acknowledgment we are expecting.
 fn has_expected_checkpoint(
-    rec: &L1BlockManifest,
+    manifest: &AsmManifest,
     expected_checkpoint: Option<&Checkpoint>,
-    params: &RollupParams,
+    _params: &RollupParams,
 ) -> bool {
-    for op in rec.txs().iter().flat_map(|tx| tx.protocol_ops()) {
-        let ProtocolOperation::Checkpoint(signed_checkpoint) = op else {
+    // Look for checkpoint ack logs in the ASM manifest
+    for log in manifest.logs() {
+        // Try to parse as SPS-52 message
+        let Some(msg) = log.try_as_msg() else {
             continue;
         };
 
-        // L1 Reader only adds a checkpoint ProtocolOperation if checkpoint signature valid.
-        let checkpoint = signed_checkpoint.checkpoint();
+        // Check if this is a checkpoint ack log
+        if msg.ty() != CHECKPOINT_ACK_ASM_LOG_TYPE_ID {
+            continue;
+        }
+
+        // Try to decode checkpoint ack data
+        let Ok(ack_data) = log.try_into_log::<CheckpointAckLogData>() else {
+            warn!(blockid = %manifest.blkid(), "failed to decode checkpoint ack log");
+            continue;
+        };
 
         // Must have expected checkpoint.
         // Can be None before first checkpoint creation, where we dont care about this.
         let Some(expected) = expected_checkpoint else {
             continue;
         };
 
-        // Must get expected checkpoint.
-        if expected.commitment() != checkpoint.commitment() {
-            warn!(got = ?checkpoint, ?expected, "got unexpected checkpoint");
-            continue;
+        // Check if the ack epoch matches our expected checkpoint
+        if ack_data.epoch().epoch() == expected.batch_info().epoch() {
+            // Found checkpoint ack for expected epoch. Should end current epoch.
```

Comment:
>Should end current epoch.

What do you mean by that? And given that we (only) return `true`, what is expected to happen next?

Reply:
When this returns `true`, the sequencer knows it's safe to end the current epoch because the previous epoch's checkpoint has been confirmed on L1.

I was trying to make the old `sequencer` work with the new ASM stuff.
I know that this is "old" code, but I don't have a better idea.